controllers/glance_controller.go

```diff
@@ -886,7 +889,39 @@ func (r *GlanceReconciler) generateServiceConfig(
 ) error {
 	labels := labels.GetLabels(instance, labels.GetGroupLabel(glance.ServiceName), serviceLabels)
 
-	ospSecret, _, err := secret.GetSecret(ctx, h, instance.Spec.Secret, instance.Namespace)
+	// create a new MariaDBAccount CR to be populated here.
+	//
+	// Note that we are only looking for a MariaDBAccount CR with a name
+	// and associated Secret, but **not** an actual mariadb username in a galera
+	// database.  The account
+	// has not yet been linked to a MariaDBDatabase, so it will only exist as a
+	// CR in the cluster.  Once linked to a MariaDBDatabase, the mariadb-operator
+	// will ensure it exists with rights for that database.
+	//
+	// This is an **interim** step, where later openstack-operator will
+	// ensure that MariaDBAccounts are created for all the services it
+	// initializes, and this block would no longer be present here.
+	// At that point the DatabaseAccount field can be set to be required.
+
```

Comment:
So, iiuc the flow here is:

1. `MariaDBAccount` CR is created via `mariadbv1.InterimCreateNewAccount`: the `mariadb-operator` watches for this kind of CRs and "translates" them into an account/database with the right permissions. A secret with the credentials is generated by mariadb and can be "consumed" by Glance;
2. We execute `mariadbv1.GetAccount`  to retrieve the associated CR, selected by `instance.Spec.DatabaseAccount`
3. Finally, we can get the secret via the regular `secret.GetSecret`, and it's selected by name via `databaseAccount.Spec.Secret`

Step 1 won't be required as long as `openstack-operator` will generate (in advance) the `MariaDBAccount` CRs for each deployed service.
As long as the CR exist, then I think we can still run the service as a standalone component w/o the openstack-operator requirement, which is good.
@zzzeek is my understanding correct?

Reply:
that's the flow with a slight extra wrinkle due to how we've so far decided we want the creation of username/password vs. the actual "database" to go:

1. `MariaDBAccount` CR is created via `mariadbv1.InterimCreateNewAccount` - ideally, this would be done by the openstack-operator itself when it first rolls out a new glance instance, or can be specified by the user creating the yaml if they want a specific username/password (this is all part of the requested use case at https://issues.redhat.com/browse/OSPRH-92 ).  but right now it's just up at the very top of the glance-operator flow as an interim architecture (this would be done for all operators).   at the moment this is also where the Glance finalizer is being added to the `MariaDBAccount` CR.
2. The `MariaDBAccount` here does not **yet** refer to a `MariaDBDatabase` - the latter is the CR that indicates that the mariadb-operator would run "CREATE DATABASE <name>" in the galera server.     Because of this, the mariadb-operator will update the status of the CR with "MariaDBDatabase not specified" and will not act upon the record yet.   (technically, you can create an account in mariadb without a specific database, then GRANT later for specific databases, however, we've designed the `MariaDBAccount` CR to always be underneath a single `MariaDBDatabase`, which corresponds more closely and simply to how we use these accounts)
3. the glance controller when it reconciles for "init" then calls `mariadbv1.NewDatabaseForAccount` , referencing the name of the `MariaDBAccount` CR as well as the name of the database, and then `db.CreateOrPatchDBByName`  - the mariadb-operator API creates a `MariaDBDatabase` CR within this run.  This is the same flow that was already here and is in every operator right now.    the Glance finalizer is added to both CRs if not already present.

The change for 3 above is that previously, the `MariaDBDatabase` would be given `osp-secret` directly (and FYI would consume `DatabasePassword` only, not `GlanceDatabasePassword`, e.g. it's a bug), whereas now there's more granularity in that there is a `MariaDBDatabase` and `MariaDBAccount` that's separate.   Multiple `MariaDBAccount` CRs can exist for a single `MariaDBDatabase` and they can also be dropped.

4. Now the mariadb-operator goes to galera and runs CREATE DATABASE for the `MariaDBDatabase` CR and CREATE USER / GRANT for the `MariaDBAccount` CR.
5. the glance operator and glance api operator continues, consuming the **username and secret** from the `MariaDBAccount`.  For accounts to be rotateable, we need the username to also be dynamic.   this username/secret is deployed into config templates to be part of the deployed openstack service.  so "username" is directly in `MariaDBAccount`, and the "secret" then refers to a `Secret` instance that has only the `DatabasePassword` field in it for this account.   Technically you can refer to `osp-secret` here since it has that key within it but it's meant to be a distinct secret
6. The glance and glance api operators as they watch the `Glance` CR will see changes in the `DatabaseAccount` name, pointing to an all new `MariaDBAccount`, which means a new username/password is being rotated in, which will occur from the outside (user intervention, or scheduled).   the controllers will need to wait for that `MariaDBAccount` to be ready, and at that point regenerate pod configs and re-deploy with new credentials.  *then* they need to *search for the old account for their glance DB based on label* and remove the `Glance` finalizer from it, so that the old account will be dropped when it's marked as deleted.

Reply:
Thank you for taking the time to explain the flow and the extra interactions I didn't catch!

> that's the flow with a slight extra wrinkle due to how we've so far decided we want the creation of username/password vs. the actual "database" to go:
> 
>     1. `MariaDBAccount` CR is created via `mariadbv1.InterimCreateNewAccount` - ideally, this would be done by the openstack-operator itself when it first rolls out a new glance instance, or can be specified by the user creating the yaml if they want a specific username/password (this is all part of the requested use case at https://issues.redhat.com/browse/OSPRH-92 ).  but right now it's just up at the very top of the glance-operator flow as an interim architecture (this would be done for all operators).   at the moment this is also where the Glance finalizer is being added to the `MariaDBAccount` CR.

Ack, makes sense, and I assume this is a simple call that will be removed/retired when we add that logic to the `openstack-operator`. I think this is a design decision that has an impact to the overall work, but I assume there's already an agreement on this part (centralize the `MariaDBAccount` creation/update at the openstack-operator level).

/cc @dprince in case there are comments or we see drawbacks with this approach.
 

> 
>     2. The `MariaDBAccount` here does not **yet** refer to a `MariaDBDatabase` - the latter is the CR that indicates that the mariadb-operator would run "CREATE DATABASE " in the galera server.     Because of this, the mariadb-operator will update the status of the CR with "MariaDBDatabase not specified" and will not act upon the record yet.   (technically, you can create an account in mariadb without a specific database, then GRANT later for specific databases, however, we've designed the `MariaDBAccount` CR to always be underneath a single `MariaDBDatabase`, which corresponds more closely and simply to how we use these accounts)

ack, so we have a 1:1 between an **active** user and database as an assumption, right? the `MariaDBAccount`, from the glance_controller perspective is a way to "request" the creation of a `user` **and** then a `MariaDBDatabase` CR is associated to an account.  Updates shouldn't be prevented because the rotation consists in rolling out a new `MariaDBAccount` for a given service and update the relation with the `MariaDBDatabase`: by updating the `Glance` CR and pointing (by `name`)  to a new `MariaDBAccount` we can implement the rotation, which consists in repeating the flow you described. Not sure we can simulate the rotation as part of the tests (probably easier via kuttl?)

> 
>     3. the glance controller when it reconciles for "init" then calls `mariadbv1.NewDatabaseForAccount` , referencing the name of the `MariaDBAccount` CR as well as the name of the database, and then `db.CreateOrPatchDBByName`  - the mariadb-operator API creates a `MariaDBDatabase` CR within this run.  This is the same flow that was already here and is in every operator right now.    the Glance finalizer is added to both CRs if not already present.

+1, right, this is already in place, we're just adding the `MariaDBAccount` CR to the picture, which reminds me something we did a while ago for the `rabbitmq-transport-url` CR that produces a `Secret` that we consume in the operator.
Should we need to add something like [1] to ensure we reference a secret that has already been created?

> 
> 
> The change for 3 above is that previously, the `MariaDBDatabase` would be given `osp-secret` directly (and FYI would consume `DatabasePassword` only, not `GlanceDatabasePassword`, e.g. it's a bug), whereas now there's more granularity in that there is a `MariaDBDatabase` and `MariaDBAccount` that's separate. Multiple `MariaDBAccount` CRs can exist for a single `MariaDBDatabase` and they can also be dropped.
> 
>     4. Now the mariadb-operator goes to galera and runs CREATE DATABASE for the `MariaDBDatabase` CR and CREATE USER / GRANT for the `MariaDBAccount` CR.
> 
>     5. the glance operator and glance api operator continues, consuming the **username and secret** from the `MariaDBAccount`.  For accounts to be rotateable, we need the username to also be dynamic.   this username/secret is deployed into config templates to be part of the deployed openstack service.  so "username" is directly in `MariaDBAccount`, and the "secret" then refers to a `Secret` instance that has only the `DatabasePassword` field in it for this account.   Technically you can refer to `osp-secret` here since it has that key within it but it's meant to be a distinct secret

Makes sense, as per your patch we're not consuming the `osp-secret` anymore, but a secret "produced" as a result of the mariadb-operator that trigger the creation logic by processing the CR.

> 
>     6. The glance and glance api operators as they watch the `Glance` CR will see changes in the `DatabaseAccount` name, pointing to an all new `MariaDBAccount`, which means a new username/password is being rotated in, which will occur from the outside (user intervention, or scheduled).   the controllers will need to wait for that `MariaDBAccount` to be ready, and at that point regenerate pod configs and re-deploy with new credentials.  _then_ they need to _search for the old account for their glance DB based on label_ and remove the `Glance` finalizer from it, so that the old account will be dropped when it's marked as deleted.

Yeah, from the `glance-operator` side a change in the CR results in a reconciliation loop that will use the new values, and in that case we already assume that an account/secret already exists.
It might be a limitation on my understanding, but in this flow I see a point where Glance should wait for an operation (the user and db creation) that should happen on the mariadb/galera side, that produces a secret that is consumed by Glance, while before we always have assumed `osp-secret` just exists and it's available. Synchronizing this operation as we've done for `transportUrls` is important, unless I overlooked part of this flow.

[1] https://github.com/openstack-k8s-operators/manila-operator/blob/main/controllers/manila_controller.go#L497

Reply:
we had a call with @stuggi and @dciabrin yesterday about TransportURL w/ rabbitMQ and how that might apply to galera.    What we sort of came up with is the answer is "sort of" - we dont really have a single, generic transport URL for galera, we have various bits that dont fit into a single URL like CAs, TLS options, but we do have a single client thing that's the local `.my.cnf` file, which we use in OOO as well for all the client side bits that are more elaborate than just host/username/password.    so we do have a plan that `.my.cnf` is going to be delivered by mariadb-operator, carried along into its destination by each operator like this glance-operator, however the specifics of the file will remain with mariadb-operator, so that changes on that end will propagate to all the controllers without them needing code changes.   we estimated that this would be similar in effect to having a transport URL set up ahead of time, just a little different.

Reply:
>  but in this flow I see a point where Glance should wait for an operation (the user and db creation) that should happen on the mariadb/galera side, that produces a secret that is consumed by Glance, while before we always have assumed osp-secret just exists and it's available. 

the osp-secret is created ahead of time but it's not useful in the Galera server until the CreateOrPatchDBByName method is called, which applies the password to a specific database that it also creates at that point.   so if the concern is "there's more waiting", there's not really any more waiting.   the MariaDBAccount will later be created up front by openstack-operator as well.

Reply:
> we had a call with @stuggi and @dciabrin yesterday about TransportURL w/ rabbitMQ and how that might apply to galera. What we sort of came up with is the answer is "sort of" - we dont really have a single, generic transport URL for galera, we have various bits that dont fit into a single URL like CAs, TLS options, but we do have a single client thing that's the local `.my.cnf` file, which we use in OOO as well for all the client side bits that are more elaborate than just host/username/password. so we do have a plan that `.my.cnf` is going to be delivered by mariadb-operator, carried along into its destination by each operator like this glance-operator, however the specifics of the file will remain with mariadb-operator, so that changes on that end will propagate to all the controllers without them needing code changes. we estimated that this would be similar in effect to having a transport URL set up ahead of time, just a little different.

right, my idea when we scheduled the call was to have the mariadbdatabase to be the owner of the mariadbaccount and then the mariadbdatabase could create a secret with the connection url and the my.cnf thats then be consumed by the operators. so basically give me DB with all required login details. for now I have submitted a PR to this operator https://github.com/openstack-k8s-operators/mariadb-operator/pull/190 to return the my.cnf

Reply:
ok,
so thank you @zzzeek for the many information and to provide a green CI: that helps to see a e2e deployment and double check we don't have any corner case/regression.
I think I have a better idea of the whole flow now.
Also, from the current CI execution I see the `MariaDBAccount` CR [1] and the resulting secrets [2].

Even though I don't like much the secret name being just `glance.yaml` (it could be like `glance-dbaccount.yaml` but this can be changed later), I think that we're going to the right direction and at the service operators level right now we're consuming the secret generated by `mariadbdatabase`.
I think what covers my question about the "wait" part is the fact we have the required info created in advance, so we should be good on that front.

Quick question looking at the CI CR(s)[1] and the secrets [2]: @gibizer does nova already implements this mechanism?
I think we should have a doc patch in [3] to describe the whole flow and the interactions so that other people know the involved CRs and how the sequence diagram of the operators interaction work.

[1] https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openstack-k8s-operators_glance-operator/426/pull-ci-openstack-k8s-operators-glance-operator-main-glance-operator-build-deploy-tempest/1755748276902039552/artifacts/glance-operator-build-deploy-tempest/openstack-k8s-operators-gather/artifacts/must-gather/quay-io-openstack-k8s-operators-openstack-must-gather-sha256-e39c97b0f556c0f58b33a035028cc0cd9a0e89f2650ef00160db8adfb85d876f/namespaces/openstack/crs/mariadbaccounts.mariadb.openstack.org/glance.yaml

[2] https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openstack-k8s-operators_glance-operator/426/pull-ci-openstack-k8s-operators-glance-operator-main-glance-operator-build-deploy-tempest/1755748276902039552/artifacts/glance-operator-build-deploy-tempest/openstack-k8s-operators-gather/artifacts/must-gather/quay-io-openstack-k8s-operators-openstack-must-gather-sha256-e39c97b0f556c0f58b33a035028cc0cd9a0e89f2650ef00160db8adfb85d876f/namespaces/openstack/secrets

[3] https://github.com/openstack-k8s-operators/docs/

Reply:
@fmount nova does not have this flow implemented yet. @zzzeek selected glance as being the first operator to get it.